flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSourceContexts.java

```diff
@@ -410,7 +402,7 @@ public WatermarkContext(
         }
 
         @Override
-        public void collect(T element) {
+        public final void collect(T element) {
             checkNotClosed();
```

Comment:
As I can see, `checkNotClosed()` was always called here on all of the code paths, so for example `processAndCollect` was calling it at least twice in the end? One time directly and the other time indirectly via this method?

Reply:
Yes. That's point 1. in the description.